pkgs/build-support/writable-dir-wrapper/default.nix

```diff
@@ -0,0 +1,186 @@
+      links? :: [{ src :: String; dst? :: String }];
+      runtimeInputs? :: [String|Derivation];
+      postLink? :: String;
+      exec :: String
+    } -> Derivation
+  ```
+
+  # Examples
+  :::{.example}
+  ## `pkgs.writableDirWrapper` usage example
+
+  ```nix
+  {
+    writableDirWrapper,
+    hello,
+  }:
+  writableDirWrapper {
+    name = "writable-hello";
+    links = [
+      {
+        src = hello;
+        dst = "hello";
+      }
+    ];
+    postLink = ''
+      echo "Hello world!" | base64 > hello.txt
+    '';
+    exec = ''
+      ./hello/bin/hello --greeting=$(<"hello.txt")
+    '';
+  }
+  ```
+
+  When built and run, this should output:
+  ```console
+  SGVsbG8gd29ybGQhCg==
+  ```
+  :::
+*/
+{
+  name,
+  version ? null,
+  path,
+  links ? [ ],
+  runtimeInputs ? [ ],
+  postLink ? "",
+  exec,
+}@args:
+let
+  args' = builtins.removeAttrs args [
+    "version"
+    "path"
+    "runtimeInputs"
+    "links"
+    "postLink"
+    "exec"
+  ];
+
+  unlinkCommand = link: ''find "${path}/${link.dst or ""}" -type l -lname "${link.src}" -delete'';
+  linkCommand = link: ''
+    mkdir -p ./${link.dst or ""}; 
+    lndir -silent ${link.src} "./${link.dst or ""}"
+  '';
+in
+writeShellApplication (
+  args'
+  // {
+    runtimeInputs = runtimeInputs ++ [ xorg.lndir ];
+
+    text =
+      ''
+        relink=false
+
+        for arg in "$@"; do
+            case "$arg" in
+              --relink-files) relink=true ;;
+              # Passthrough all other args
+              *) args+=("$arg") ;;
+            esac
+        done
+
+        versionFile="${path}/.${name}.version"
+
+        if [[ "$relink" = "true" ]]; then
+          echo "Manually relinking files"
+        elif [[ ! -d "${path}" ]]; then
+          echo "Directory ${path} not found. Linking files..."
+          relink=true
+      ''
+      + lib.optionalString (version != null) ''
+        elif [[ ! -e "$versionFile" ]]; then
+          echo "Version file not found. Relinking files just to be safe..."
+          relink=true
+        elif [[ "${version}" != "$(<"$versionFile")" ]]; then
+          echo "Current version is not the same as the newest version (${version}). Relinking files"
+          relink=true
+      ''
+      + ''
+        fi
+
+        if [[ "$relink" = "true" ]]; then
+          mkdir -p "${path}"
+          if [[ -d "${path}" ]]; then
+            echo "Removing existing file links"
+            ${lib.concatMapStringsSep "\n" unlinkCommand links}
+          fi
+
+          echo "Linking files"
+          ${lib.optionalString (version != null) ''echo "${version}" > "$versionFile"''}
+
+          # In case that postLink needs to remove some links to avoid conflict,
+          # we first link everything into a temporary directory then move over
+          (
+            cd "$(mktemp -d)"
+            ${lib.concatMapStringsSep "\n" linkCommand links}
+            ${postLink}
+            cp -r ./* -t "${path}"
+          )
+        fi
```

Comment:
I'd extract this to a function, then call it whenever you're setting relink to true, rather than having a boolean.